federation/pkg/federation-controller/replicaset/planner/planer.go

```diff
@@ -0,0 +1,144 @@
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package planer
+
+import (
+	"sort"
+
+	fed_api "k8s.io/kubernetes/federation/apis/federation"
+)
+
+// Planer decides how many out of the given replicas should be placed in each of the
+// federated clusters.
+type Planer struct {
+	preferences *fed_api.FederatedReplicaSetPreferences
+}
+
+type namedLocalReplicaSetPreferences struct {
+	clusterName string
+	fed_api.LocalReplicaSetPreferences
+}
+
+type byWeight []*namedLocalReplicaSetPreferences
+
+func (a byWeight) Len() int      { return len(a) }
+func (a byWeight) Swap(i, j int) { a[i], a[j] = a[j], a[i] }
+
+// Preferences are sorted according by decreasing weight and increasing clusterName.
+func (a byWeight) Less(i, j int) bool {
+	return (a[i].Weight > a[j].Weight) || (a[i].Weight == a[j].Weight && a[i].clusterName < a[j].clusterName)
+}
+
+func NewPlaner(preferences *fed_api.FederatedReplicaSetPreferences) *Planer {
+	return &Planer{
+		preferences: preferences,
+	}
+}
+
+// Distribute the desired number of replicas among the given cluster according to the planner preferences.
+func (p *Planer) Plan(replicasToDistribute int64, availableClusters []string) map[string]int64 {
+	preferences := make([]*namedLocalReplicaSetPreferences, 0, len(availableClusters))
+
+	named := func(name string, pref fed_api.LocalReplicaSetPreferences) *namedLocalReplicaSetPreferences {
+		return &namedLocalReplicaSetPreferences{
+			clusterName:                name,
+			LocalReplicaSetPreferences: pref,
+		}
+	}
+
+	for _, cluster := range availableClusters {
+		if localRSP, found := p.preferences.Clusters[cluster]; found {
+			preferences = append(preferences, named(cluster, localRSP))
+		} else {
+			if localRSP, found := p.preferences.Clusters["*"]; found {
+				preferences = append(preferences, named(cluster, localRSP))
+			} else {
+				preferences = append(preferences, named(cluster, fed_api.LocalReplicaSetPreferences{Weight: 0}))
+			}
+		}
+	}
+	sort.Sort(byWeight(preferences))
+
+	remainingReplicas := replicasToDistribute
+	plan := make(map[string]int64, len(preferences))
+
+	// Assign each cluster the minimum number replica it requested.
+	for _, preference := range preferences {
+		min := minInt64(preference.MinReplicas, remainingReplicas)
+		remainingReplicas -= min
+		plan[preference.clusterName] = min
+	}
+
+	modified := true
+
+	// It is possible single pass of the loop is not enough to distribue all replicas among clusters due
+	// to weight, max and rounding corner cases. In such case we iterate until either
```

Comment:
suggestion: It might be possible to round strictly upwards, and then trim the excess replicas?  Might allow you to d a single pass?